src/ray/util/process.cc

```diff
@@ -0,0 +1,325 @@
+  pid_t pid_;
+  intptr_t fd_;
+
+ public:
+  ~ProcessFD();
+  ProcessFD();
+  ProcessFD(pid_t pid, intptr_t fd = -1);
+  ProcessFD(const ProcessFD &other);
+  ProcessFD(ProcessFD &&other);
+  ProcessFD &operator=(const ProcessFD &other);
+  ProcessFD &operator=(ProcessFD &&other);
+  intptr_t CloneFD() const;
+  void CloseFD();
+  intptr_t GetFD() const;
+  pid_t GetId() const;
+
+  // Fork + exec combo. Returns -1 for the PID on failure.
+  static ProcessFD spawnvp(const char *argv[], std::error_code &ec) {
+    ec = std::error_code();
+    intptr_t fd;
+    pid_t pid;
+#ifdef _WIN32
+    fd = _spawnvp(P_NOWAIT, argv[0], argv);
+    if (fd != -1) {
+      pid = static_cast<pid_t>(GetProcessId(reinterpret_cast<HANDLE>(fd)));
+      if (pid == 0) {
+        pid = -1;
+      }
+    } else {
+      pid = -1;
+    }
+    if (pid == -1) {
+      ec = std::error_code(GetLastError(), std::system_category());
+    }
+#else
+    // TODO(mehrdadn): Use clone() on Linux or posix_spawnp() on Mac to avoid duplicating
+    // file descriptors into the child process, as that can be problematic.
+    pid = fork();
+    if (pid == 0) {
+      // Child process case. Reset the SIGCHLD handler for the worker.
+      signal(SIGCHLD, SIG_DFL);
+      if (execvp(argv[0], const_cast<char *const *>(argv)) == -1) {
+        pid = -1;
+        abort();  // fork() succeeded but exec() failed, so abort the child
+      }
+    }
+    if (pid == -1) {
+      ec = std::error_code(errno, std::system_category());
+    }
+    // TODO(mehrdadn): This would be a good place to open a descriptor later
+    fd = -1;
+#endif
+    return ProcessFD(pid, fd);
+  }
+};
+
+ProcessFD::~ProcessFD() {
+  if (fd_ != -1) {
+#ifdef _WIN32
+    CloseHandle(reinterpret_cast<HANDLE>(fd_));
+#else
+    close(static_cast<int>(fd_));
+#endif
+  }
+}
+
+ProcessFD::ProcessFD() : pid_(-1), fd_(-1) {}
+
+ProcessFD::ProcessFD(pid_t pid, intptr_t fd) : pid_(pid), fd_(fd) {
+  if (pid != -1) {
+    bool process_does_not_exist = false;
+    std::error_code error;
+#ifdef _WIN32
+    if (fd == -1) {
+      BOOL inheritable = FALSE;
+      DWORD permissions = MAXIMUM_ALLOWED;
+      HANDLE handle = OpenProcess(permissions, inheritable, static_cast<DWORD>(pid));
+      if (handle) {
+        fd_ = reinterpret_cast<intptr_t>(handle);
+      } else {
+        DWORD error_code = GetLastError();
+        error = std::error_code(error_code, std::system_category());
+        if (error_code == ERROR_INVALID_PARAMETER) {
+          process_does_not_exist = true;
+        }
+      }
+    } else {
+      RAY_CHECK(pid == GetProcessId(reinterpret_cast<HANDLE>(fd)));
+    }
+#else
+    if (kill(pid, 0) == -1 && errno == ESRCH) {
+      process_does_not_exist = true;
+    }
+#endif
+    // Don't verify anything if the PID is too high, since that's used for testing
+    if (pid < PID_MAX_LIMIT) {
+      if (process_does_not_exist) {
+        RAY_LOG(FATAL) << "Process " << pid << " does not appear to exist. "
```

Comment:
Let's change this to "Internal error: Process {pid} owned by the raylet does not exist.". "Internal error" indicates that this should not happen and it is a logic error. On the off-chance that a user sees this error, this will be less intimidating to them.

If you want to you can add a comment like this above the RAY_LOG line:
// The following error should not happen: For owned processes, the PIDs are valid until the raylet calls wait() on them, even if the process already exited. To make sure the PID stays valid, SIGCHLD should NOT be suppressed.